Avoid leaking of tracked memory for async logging and text_log

azat opened the pull request:
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Avoid leaking of tracked memory for async logging (can have a significant drift, for 10 hours, ~100GiB) and text_log (almost same drift is possible).

Fixes: https://github.com/ClickHouse/ClickHouse/issues/82036

azat committed:
Avoid leaking of tracked memory with async logging

Note that we can remove blocker for SystemLogBase::push() since it is
done from logSplit() method

Algunenano reviewed (approved):
LGTM. A few of comments:
* Could you please leave a comment explaining why it's done. It's obvious to me now, but probably not to everyone and not me in a few months.
* I think we can remove https://github.com/ClickHouse/ClickHouse/blob/4117df841f4bed5fb4b6418e455030a3e2b413e0/src/Loggers/OwnSplitChannel.cpp#L135 since a) it's already covered here and b) it's also covered by `SystemLogQueue<LogElement>::push`.
* Could this have been an issue inside sync logging when calling `pushExtendedMessageToInternalTCPTextLogQueue`? It seems it might grow the queue, allocating extra memory that's not freed but it's tied to `CurrentThread` so probably accounting was kept correctly.
* This makes me wonder what happens in logging threads if an exception is thrown when logging, inside `while (is_open)`. It seems the thread might stop and never recover. I'll investigate further

azat commented on src/Loggers/OwnSplitChannel.cpp:
```diff
@@ -395,7 +396,7 @@ void OwnAsyncSplitChannel::log(const Poco::Message & msg)
 void OwnAsyncSplitChannel::log(Poco::Message && msg)
 {
-    LockMemoryExceptionInThread lock_memory_tracker(VariableContext::Global);
+    MemoryTrackerBlockerInThread block_memory_tracker(VariableContext::Global);
```
But we move the message there - https://github.com/ClickHouse/ClickHouse/blob/7891d941df3e69cf2eca90bae420320bedce3d6e/src/Loggers/OwnSplitChannel.cpp#L293-L294

Algunenano replied:
Some approaches:
* Go back to copying the message
* Block memory tracker earlier (https://github.com/ClickHouse/ClickHouse/blob/bec82f8bacbeeab927da62ec4a4da8ac8b19fe8c/src/Common/logger_useful.h#L124)

azat replied:
Correct. Now it is in `logger_useful.h`

azat committed:
Add missing fwd decl of ExtendedLogMessage

azat committed:
Add missing trace_type's into trace_log documentation

azat committed:
Add a comment for MemoryTrackerDebugBlockerInThread

azat committed:
Update peak for global memory tracker even if it is blocked

azat committed:
Move MemoryTrackerDebugBlockerInThread into SystemLogQueue::push()

azat commented:
>Could this have been an issue inside sync logging when calling pushExtendedMessageToInternalTCPTextLogQueue? It seems it might grow the queue, allocating extra memory that's not freed but it's tied to CurrentThread so probably accounting was kept correctly.

Actually I think for this case it should be accounted properly, since `TCPHandler` (and `HTTPHandler`) has `CurrentThread::QueryScope` which creates initial thread group for a query

azat committed:
Properly fix memory tracking for logs

azat committed:
Remove MemoryTrackerDebugBlockerInThread from SystemLogQueue::add()

This is likely redundant since we have it in push() which does big
allocations, add() should not exceed 16 MiBs

azat committed:
Get back MemoryTrackerBlockerInThread into the SystemLogBase

azat committed:
Fix coding style in SystemLogBase

azat commented:
>[Integration tests (amd_tsan, 2/6)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=87584&sha=e1464e87807b4e09159c70abcf65cb979259946e&name_0=PR&name_1=Integration%20tests%20%28amd_tsan%2C%202%2F6%29)

- `test_storage_kafka` - https://github.com/ClickHouse/ClickHouse/issues/87654

>[Stateless tests (amd_binary, ParallelReplicas, s3 storage, parallel)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=87584&sha=e1464e87807b4e09159c70abcf65cb979259946e&name_0=PR&name_1=Stateless%20tests%20%28amd_binary%2C%20ParallelReplicas%2C%20s3%20storage%2C%20parallel%29)

- https://github.com/ClickHouse/ClickHouse/issues/87653

>[Integration tests (arm_binary, distributed plan, 4/4)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=87584&sha=e1464e87807b4e09159c70abcf65cb979259946e&name_0=PR&name_1=Integration%20tests%20%28arm_binary%2C%20distributed%20plan%2C%204%2F4%29)

- `test_keeper_memory_soft_limit/test.py::test_soft_limit_create` - **will take a look**

azat committed:
tests: fix test_keeper_memory_soft_limit flakiness

CI: https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=87584&sha=e1464e87807b4e09159c70abcf65cb979259946e&name_0=PR&name_1=Integration%20tests%20%28arm_binary%2C%20distributed%20plan%2C%204%2F4%29

azat committed:
tests: cleanup test_keeper_memory_soft_limit

azat commented:
It has been battle tested by @nikitamikhaylov on a long running `BACKUP`

azat commented:
CI:

- https://github.com/ClickHouse/ClickHouse/issues/87653
- docker hub issues

```
E               Error response from daemon: Get "https://registry-1.docker.io/v2/": net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
```

azat committed:
test: fix test_keeper_memory_soft_limit
